src/main/java/org/opensearch/knn/index/query/filtered/FilteredKNNIterator.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query.filtered;
+
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.Setter;
+import org.apache.lucene.index.BinaryDocValues;
+import org.opensearch.knn.index.SpaceType;
+
+import java.io.IOException;
+
+/**
+ * Inspired by DiversifyingChildrenFloatKnnVectorQuery in lucene
+ * https://github.com/apache/lucene/blob/7b8aece125aabff2823626d5b939abf4747f63a7/lucene/join/src/java/org/apache/lucene/search/join/DiversifyingChildrenFloatKnnVectorQuery.java#L162
+ *
+ * The class is used in KNNWeight to score filtered KNN field by iterating filterIdsArray.
+ */
+@Getter(AccessLevel.PROTECTED)
+@Setter(AccessLevel.PROTECTED)
+public abstract class FilteredKNNIterator<T extends KNNQueryVector> {
+    // Array of doc ids to iterate
+    private final int[] filterIdsArray;
+    private float currentScore = Float.NEGATIVE_INFINITY;
+    private final T queryVector;
+    private final BinaryDocValues values;
+    private final SpaceType spaceType;
+    private int currentPos = 0;
+
+    public FilteredKNNIterator(final int[] filterIdsArray, final T queryVector, final BinaryDocValues values, final SpaceType spaceType) {
+        this.filterIdsArray = filterIdsArray;
+        this.queryVector = queryVector;
+        this.values = values;
+        this.spaceType = spaceType;
+    }
+
+    /**
+     * Advance to the next doc and update score
```

Comment:
can we add more details on why and how score is updated?

Reply:
ack